### PR TITLE
coala: Add filter-by option for json and format

### PR DESCRIPTION
**Pull request by shreyans800755**
**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [x] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [x] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

**Comment by shreyans800755** [comment]
@Makman2 please review this PR.

File: coalib/coala_modes.py [hunk 1]
```diff
@@ -46,34 +46,17 @@ def mode_json(args, debug=False):
...
-                results.append(bear)
-        except BaseException as exception:  # pylint: disable=broad-except
-            return get_exitcode(exception)
-    else:
-        results, exitcode, _ = run_coala(args=args, debug=debug)
 
-    retval = {'bears': results} if args.show_bears else {'results': results}
+    results, exitcode, _ = run_coala(args=args, debug=debug)
+
+    retval = {'results': results}
```
Comment (Makman2) [inline coalib/coala_modes.py]: do we change json output with that? If we have `show-bears` enabled, it's put into the `results` field in JSON, not into `bears` any more.

Comment (shreyans800755) [reply coalib/coala_modes.py]: No. If `show-bears` is specified along with json mode, execution flow won't come to this file. Have a look at https://github.com/coala/coala/pull/4726/files/fa62b7ba0d032b0d65a8c25c57f121b738a04202#diff-aed82cdd504caf613603b250c092f460L56. I've shifted this condition below `show-bears` condition in `coala.py` file.

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -407,6 +407,16 @@ def print_results_formatted(log_printer,
...
 
 
+def print_bears_formatted(bears, format=None):
+    format_str = format or ('name:{name}:can_detect:{can_detect}:'
+                            'can_fix:{can_fix}:description:{description}')
+    print('\n\n'.join(format_str.format(name=bear.name,
+                                        can_detect=bear.CAN_DETECT,
+                                        can_fix=bear.CAN_FIX,
+                                        description=bear.get_metadata().desc)
+                      for bear in bears))
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: is it possible to have 2 commits, one for implementing `filter-by` for `format` and one for `json`?

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -922,27 +933,58 @@ def print_bears(bears,
...
+                              cls=JSONEncoder,
+                              sort_keys=True,
+                              indent=2,
+                              separators=(',', ': '))
+            else:
+                print(json.dumps(json_output,
+                                 cls=JSONEncoder,
+                                 sort_keys=True,
+                                 indent=2,
+                                 separators=(',', ': ')))
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: can't you use `json.dumps` once and just write it to a file? This deduplicates the call ^^

File: coalib/output/ConsoleInteraction.py [hunk 4]
```diff
@@ -922,27 +933,58 @@ def print_bears(bears,
...
     if not bears:
         console_printer.print('No bears to show. Did you forget to install '
                               'the `coala-bears` package? Try `pip3 install '
                               'coala-bears`.')
         return
 
+    results = []
+    if args and args.json:
+        from coalib.output.JSONEncoder import create_json_encoder
+        JSONEncoder = create_json_encoder(use_relpath=args.relpath)
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: can't you create the JSONEncoder below? Then you need one if less

**Comment by shreyans800755** [comment]
@Makman2 Pushed the changes you requested. I can split the commit if its really required.

File: coalib/output/ConsoleInteraction.py [hunk 5]
```diff
@@ -922,27 +933,58 @@ def print_bears(bears,
...
+            show_bear(bear,
+                      show_description,
+                      show_params,
+                      console_printer)
+
+    if args:
+        if args.json:
+            json_output = {'bears': results}
+            import json
+            if args.output:
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: what if output is provided for `format`?

File: coalib/output/ConsoleInteraction.py [hunk 6]
```diff
@@ -922,27 +933,58 @@ def print_bears(bears,
...
+                              indent=2,
+                              separators=(',', ': '))
+            else:
+                print(json.dumps(json_output,
+                                 cls=JSONEncoder,
+                                 sort_keys=True,
+                                 indent=2,
+                                 separators=(',', ': ')))
+        elif args.format:
+            print_bears_formatted(results)
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: maybe put here
```
for bear in results:
    show_bear(...)
```
from above, to make this more readable in code. So you have the bear-gathering phase and the printing phase. However you need return statements then for each case from above, e.g.:
```
print_bears_formatted(...)
return
```

Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: Yeah makes sense. I'm kinda checking `args and (args.json or args.format)` condition for all the bears individually, which is unnecessary.

File: coalib/output/ConsoleInteraction.py (hunk 5 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: As per help section,
```
-o, --output          write results to the given file (must be called with
                        --json)
```

IMO, if we want this option to be used with other entrypoints in the similar context, we should file different issue.

File: coalib/output/ConsoleInteraction.py (hunk 6 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: I can't see scenario of error over here. So, by default coala.py returns 0 after calling show_bears function from ConsoleInteraction module. I haven't changed that part.

Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: > I can't see scenario of error over here. So, by default coala.py returns 0 after calling show_bears function from ConsoleInteraction module. I haven't changed that part.

What do you mean?

Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: https://github.com/coala/coala/pull/4726/commits/f039489890654bc7396f7a482c1c62fcc96bf571#diff-aed82cdd504caf613603b250c092f460L88 returns success - zero after calling `show_bears` function with appropriate entry point through `args` param. So, we don't need to return any error code from `show_bears` function.

**Comment by shreyans800755** [comment]
@gitmate-bot rebase

File: coalib/output/ConsoleInteraction.py [hunk 7]
```diff
@@ -922,27 +933,49 @@ def print_bears(bears,
...
+        JSONEncoder = create_json_encoder(use_relpath=args.relpath)
+        json_output = {'bears': results}
+        import json
+        json_formatted_output = json.dumps(json_output,
+                                           cls=JSONEncoder,
+                                           sort_keys=True,
+                                           indent=2,
+                                           separators=(',', ': '))
+        if args.output:
+            filename = str(args.output[0])
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: `args.output[0]`? Why not `args.output`?

File: coalib/output/ConsoleInteraction.py [hunk 8]
```diff
@@ -922,27 +933,49 @@ def print_bears(bears,
...
+            with open(filename, 'w+') as fp:
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: `w+`, so write and append? That would corrupt existing json files, the parent element has to be either an item or an array in json syntax.

File: coalib/output/ConsoleInteraction.py [hunk 9]
```diff
@@ -922,27 +933,49 @@ def print_bears(bears,
...
+                fp.write(json_formatted_output)
+        else:
+            print(json_formatted_output)
+    elif args and args.format:
+        print_bears_formatted(results)
```
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: hmm so we don't support `args.output`... I won't change this commit that much it's still very good, however we have to streamline the functionality of different output formats and arguments like `args.output`.

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: `args.output` is list of file names given as args. It is used in similar manner in other places (https://github.com/coala/coala/pull/4726/commits/a581d099b1d8b8dab3a910f3ef2e2a974e67338f#diff-a297897b536e74044ca36e47697b169dL83) as well.

File: coalib/output/ConsoleInteraction.py (hunk 8 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: It is same as https://github.com/coala/coala/pull/4726/commits/a581d099b1d8b8dab3a910f3ef2e2a974e67338f#diff-a297897b536e74044ca36e47697b169dL84. @tulphoon has made last changes there. Should I add new commit to correct it at both the locations ?

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: @Makman2

Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: Shouldn't we then iterate over each file and write to each of them??

File: coalib/output/ConsoleInteraction.py (hunk 8 above)
Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: yeah sounds good :+1:

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: True. But do we really need an option like `output` ?
If user is using cli, there's always an option to use Linux i/o redirectives. It's actually easy to remember and shorter to write.

Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: though we should still support it for now (with multiple files) and discussing later the usefulness of `--output`

Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: Ah and you should adapt tests to use multiple output files to test that properly ;)

Comment (shreyans800755) [reply coalib/output/ConsoleInteraction.py]: Ah! Only one argument is supported after `--output` using the parameter `nargs=1` while building parser. But it is stored in a list by default irrespective of number of args.
Do we still want to implement this for multiple files ?

**Comment by shreyans800755** [comment]
@Makman2 
Any idea why appveyor is failing ?

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: Ah because of `nargs`... I think if someone omits `nargs=1` in the parameter definition altogether, it requires by default a single value and doesn't use an array to store it. However, this is not job of this PR to change that, so keep it ;)

Comment (Makman2) [reply coalib/output/ConsoleInteraction.py]: Maybe add a comment then? Not sure...

**Comment by Makman2** [comment]
Ugh damn upstream break... Got this error on another PR too...

**Comment by Makman2** [comment]
1. see suggestion of gitmate ;)
2. you should maybe again use `coala` as a commit tag
3. Add the reason why you are now truncating the files before instead of appending, I think this is important (especially without an issue)

File: coalib/output/ConsoleInteraction.py (hunk 7 above)
Comment (Makman2) [inline coalib/output/ConsoleInteraction.py]: Isn't `args.output[0]` already a string? Then you don't need to invoke `str`

File: tests/coalaJSONTest.py [hunk 10]
```diff
@@ -150,3 +151,17 @@ def test_output_file(self):
...
+                coala.main, 'coala', '--json', '-B', '--filter-by', 'language',
+                'java', '-I', '--output', 'bears.json')
+
+        with open('bears.json') as fp:
+            data = json.load(fp)
+
+        self.assertEqual(retval, 0)
+        self.assertEqual(len(data['bears']), 2)
+        self.assertFalse(stderr)
+        os.remove('bears.json')
```
Comment (Makman2) [inline tests/coalaJSONTest.py]: remove the file immediately once you've loaded contents, then the file doesn't remain on test fails (which would cause then test fails, because fo the append mode!)

File: tests/coalaJSONTest.py [hunk 11]
```diff
@@ -150,7 +151,6 @@ def test_output_file(self):
         self.assertFalse(retval2)
         self.assertFalse(stdout2)
         self.assertFalse(stderr2)
-        os.remove('file.json')
```
Comment (Makman2) [inline tests/coalaJSONTest.py]: this change should go into a separate commit

File: tests/coalaJSONTest.py [hunk 12]
```diff
@@ -160,8 +160,8 @@ def test_show_language_bears_output_file(self):
 
         with open('bears.json') as fp:
             data = json.load(fp)
+        os.remove('bears.json')
 
         self.assertEqual(retval, 0)
         self.assertEqual(len(data['bears']), 2)
         self.assertFalse(stderr)
-        os.remove('bears.json')
```
Comment (Makman2) [inline tests/coalaJSONTest.py]: and this change should be in the previous commit introducing the test

File: coalib/coala_modes.py [hunk 13]
```diff
@@ -64,7 +64,7 @@ def mode_json(args, debug=False):
 
     if args.output:
         filename = str(args.output[0])
-        with open(filename, 'w+') as fp:
+        with open(filename, 'w') as fp:
```
Comment (Makman2) [inline coalib/coala_modes.py]: actually, do we wanna test that? Simply write to the same file twice, and check if content was replaced and not appended^^

Comment (shreyans800755) [reply coalib/coala_modes.py]: IMO, it doesn't make sense to add tests for this as its python internals. Just for safety, I've tested it locally and seems to be working fine.

Comment (Makman2) [reply coalib/coala_modes.py]: >  it doesn't make sense to add tests for this as its python internals.

It definitely reflects on the user. That's why we should test it.
However, let's open an issue for that afterwards, imo not so important right now as it's currently not officially standardized I believe whether we append or overwrite output files.

**Comment by Makman2** [comment]
For the commit `Delete output file immediately after loading data` I definitely want to see the tag `coalaJSONTest`. So `coalaJSONTest: Delete output file immediately` (imo the context of tests is important here)

**Comment by Makman2** [comment]
ack 8492fe9 50d9e32 3458c29

**Comment by Makman2** [comment]
@rultor merge
